PayloadTrie.parseText method inconsistencies

omarshibli opened the pull request:
Hi,

I've noticed parseText method inconsistencies if used with `onlyWholeWords` or `onlyWholeWordsWhiteSpaceSeparated` options.

Here is a test to show the issue:
```java
    @Test
    public void testWholeWords() {
        PayloadTrie<String> trie = PayloadTrie.<String>builder().addKeyword("foo", "bar").onlyWholeWords().build();

        // access via org.ahocorasick.trie.PayloadTrie.parseText(java.lang.CharSequence)
        Collection<PayloadEmit<String>> result1 = trie.parseText("foobar");


        // access via org.ahocorasick.trie.PayloadTrie.parseText(java.lang.CharSequence, org.ahocorasick.trie.handler.PayloadEmitHandler<T>)
        Collection<PayloadEmit<String>> result2 = new LinkedList<>();
        trie.parseText("foobar", result2::add);

        assertEquals(result1, result2);
    }
```

Attached proposed changes to make both these use cases consistent.

Note that `ignoreOverlaps` option is still inconsistent in both usages, but it's more complicated to solve and I prefer to keep it separated from this PR.

# Issue

https://github.com/robert-bor/aho-corasick/issues/85

omarshibli commented:
I did some benchmarks, this new code improves `parseText` method, observed improvements:
- for matching strings 
  - up to **10% improvement** for no match case
  - up to **13% improvement** for a match case
- for matching words ( 
  - up to **15% improvement** for no match case
  - up to **37% improvement** for a match case

Here is [benchmark code for reference](https://github.com/omarshibli/aho-corasick/blob/benchmark/src/main/java/org/ahocorasick/benchmark/BenchmarkPayloadTrie.java), [result files here](https://github.com/omarshibli/aho-corasick/tree/benchmark/benchmark).

DaveJarvis commented on src/test/java/org/ahocorasick/interval/IntervalTest.java:
```diff
@@ -54,4 +54,14 @@ public void comparable() {
     }
 
+    @Test
+    public void checkToString() {
```
Let's migrate the naming schemes for the unit test methods to follow Roy Osherove's recommendation: https://osherove.com/blog/2005/4/3/naming-standards-for-unit-tests.html

With one minor modification to adhere to Google's Java style guide conventions: prefix the method name with `test_`.

DaveJarvis commented on src/test/java/org/ahocorasick/trie/StateTest.java:
```diff
@@ -19,6 +22,52 @@ public void constructSequenceOfCharacters() {
+
+    @Test
+    public void getStates() {
+        State rootState = new State();
```
When possible, please use a `final` modifier, even for local member variables.

DaveJarvis commented on src/test/java/org/ahocorasick/trie/StateTest.java:
```diff
@@ -19,6 +22,52 @@ public void constructSequenceOfCharacters() {
+        rootState.setFailure(failureState);
+
+        assertEquals(failureState, rootState.failure());
+
```
Style nit: No newline before the closing brace inside methods.

omarshibli replied:
👍

DaveJarvis commented on src/test/java/org/ahocorasick/trie/StateTest.java:
```diff
@@ -19,6 +22,52 @@ public void constructSequenceOfCharacters() {
         currentState = currentState.nextState('c');
         assertEquals(3, currentState.getDepth());
+        currentState = currentState.nextState('F');
+        assertNull(currentState);
```
Is there a way to avoid `null` here? Perhaps the `nextState` should return `Optional`?

omarshibli replied:
I don't think, it's a good idea, it's an internal API, the reason I added all these extra unit tests, just for the sake of improving the code coverage.
Code coverage has dropped probably because I removed `ListElementRemoval.java` and its unit test altogether.

DaveJarvis commented on src/test/java/org/ahocorasick/trie/PayloadTrieTest.java:
```diff
@@ -449,8 +449,9 @@ public void testLargeString() {
+    // @see https://github.com/robert-bor/aho-corasick/issues/85
     @Test
-    public void testWholeWords() {
+    public void wholeWords() {
```
Please revise to use Roy's naming scheme: https://osherove.com/blog/2005/4/3/naming-standards-for-unit-tests.html, prefixed with `test_`.

omarshibli replied:
👍

DaveJarvis commented on src/test/java/org/ahocorasick/trie/PayloadTrieTest.java:
```diff
@@ -461,6 +462,24 @@ public void testWholeWords() {
+
+        // access via PayloadTrie.parseText(CharSequence)
+        Collection<PayloadEmit<String>> result1 = trie.parseText("foo#bar");
+
```
Style nit: Prefer to avoid double-blank lines.

omarshibli replied:
👍

DaveJarvis commented on src/test/java/org/ahocorasick/trie/PayloadTrieTest.java:
```diff
@@ -461,6 +462,24 @@ public void testWholeWords() {
+        assertEquals(result1, result2);
+    }
+
+    // @see https://github.com/robert-bor/aho-corasick/issues/85
```
A couple of items of note:

1. Prefer to not reference GitHub issues (if we migrate to a different host, the comment goes stale).
1. Prefer to use `/** ... */` style comment for methods; it's okay to paste the relevant parts of the issue into the source code comment.

omarshibli replied:
same thing as below, I saw it's pretty common in this unit test class to add links to the related issues. I think it's a good for reference.

DaveJarvis commented on src/test/java/org/ahocorasick/trie/PayloadTrieTest.java:
```diff
@@ -449,8 +449,9 @@ public void testLargeString() {
         assertEquals(textSize / interval, emits.size());
     }
 
+    // @see https://github.com/robert-bor/aho-corasick/issues/85
```
See note about external links to GitHub.

omarshibli replied:
where is the note? I saw it's pretty common in this unit test class to add links to the related issues. I think it's a good for reference.

DaveJarvis reviewed (changes requested):
Thank you so much for the PR and the performance improvements. There are a few changes I'd like to see get started in the code base, including unit test naming standards and use of `final` modifier. The remaining changes look good.

The code coverage looks like it has dropped a bit. Can you see a way to raise it back to 85%?

omarshibli commented:
Code coverage probably dropped because I removed `ListElementRemoval` class and its unit test. I added `StateTest` and `IntervalTest` to increase the code coverage, but it seems this is not enough..

DaveJarvis commented:
@omarshibli, I made some changes to the main branch. Formatting, fixing IDE warnings, introduced some lambda expressions, added source code comments, etc. If you are willing to resolve the conflicts, I'll accept the PR. Thank you so much!

omarshibli commented:
@DaveJarvis sure, will resolve conflicts and go over your comments again. thanks.

omarshibli commented:
@DaveJarvis updated PR based on your comments, please review.
